lib/Interpreter/BackendPasses.cpp

```diff
@@ -69,6 +70,52 @@ namespace {
 
 char KeepLocalGVPass::ID = 0;
 
+namespace {
+
+  // Add a suffix to the CUDA module ctor/dtor to generate an unique name.
+  // It's necessary for lazy compilation. Without suffix, cling can't
+  // differentiate the ctor/dtor of the first module from the ctor/dtor of the
+  // following modules.
+  class UnqiueCUDACtorDtorName : public ModulePass {
+    static char ID;
+
+    bool runOnGlobal(GlobalValue& GV, StringRef & ModuleName){
+      if(GV.hasName() && (GV.getName() == "__cuda_module_ctor"
+          || GV.getName() == "__cuda_module_dtor") ){
+        llvm::SmallString<128> NewFunctionName;
+        NewFunctionName.append(GV.getName());
+        NewFunctionName.append("_");
+        NewFunctionName.append(ModuleName);
+
+        for (size_t i = 0; i < NewFunctionName.size(); ++i) {
+          // Replace everything that's not [a-zA-Z0-9._] with a _. This set happens
+          // to be the set of C preprocessing numbers.
+          if (!isPreprocessingNumberBody(NewFunctionName[i]))
+            NewFunctionName[i] = '_';
+        }
+
+        GV.setName(NewFunctionName);
+
+        return true;
+      }
+
+      return false;
+    }
+
+  public:
+    UnqiueCUDACtorDtorName() : ModulePass(ID) {}
+
+    bool runOnModule(Module &M) override {
+      bool ret = false;
+      StringRef ModuleName = M.getName();
+      for (auto &&F: M)
```

Comment:
`F` should be of type `llvm::Function`. Should `runOnGlobal` be called `runOnFunction` (and take a `Function`)?

Reply:
It's true, F is a llvm::Function object. I could use llvm::Function instead llvm::GlobalValue. Basically, `__cuda_module_ctor` and `__cuda_module_dtor` are just functions. But in the context of the program, they get the meaning of a global ctor and dtor.

I think, function is better, because we don't need to cast the object and in this moment of the compiler pipeline, it is just a simple function and no global ctor or dtor.

Reply:
> they get the meaning of a global ctor and dtor

Well, any `llvm::Function` is a `llvm::Global`... These are all just symbols. Not sure I understand your argument, but I believe we agree on the conclusion!